src/jdk.net/aix/native/libextnet/AIXSocketOptions.c

```diff
@@ -0,0 +1,265 @@
+ * questions.
+ */
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <sys/types.h>
+#include <string.h>
+#include <errno.h>
+#include <unistd.h>
+
+#include <jni.h>
+#include <netinet/tcp.h>
+#include <netinet/in.h>
+#include "jni_util.h"
+#include "jdk_net_AIXSocketOptions.h"
+
+
+static void handleError(JNIEnv *env, jint rv, const char *errmsg) {
+    if (rv < 0) {
+        if (errno == ENOPROTOOPT) {
+            JNU_ThrowByName(env, "java/lang/UnsupportedOperationException",
+                    "unsupported socket option");
+        } else {
+            JNU_ThrowByNameWithLastError(env, "java/net/SocketException", errmsg);
+        }
+    }
+}
+
+static jint socketOptionSupported(jint level, jint optname) {
+    jint one = 1;
+    jint rv, s;
+    socklen_t sz = sizeof (one);
+    /* First try IPv6; fall back to IPv4. */
+    s = socket(PF_INET6, SOCK_STREAM, IPPROTO_TCP);
+    if (s < 0) {
+        if (errno == EPFNOSUPPORT || errno == EAFNOSUPPORT) {
+            s = socket(PF_INET, SOCK_STREAM, IPPROTO_TCP);
+        }
+        if (s < 0) {
+            return 0;
+        }
+    }
+    rv = getsockopt(s, level, optname, (void *) &one, &sz);
+    if (rv != 0 && errno == ENOPROTOOPT) {
+        rv = 0;
+    } else {
+        rv = 1;
+    }
+    close(s);
+    return rv;
+}
+
+/*
+ * Declare library specific JNI_Onload entry if static build
+ */
+DEF_STATIC_JNI_OnLoad
+
+/*
+ * Class:     jdk_net_AIXSocketOptions
+ * Method:    setQuickAck
+ * Signature: (II)V
+ */
+JNIEXPORT void JNICALL Java_jdk_net_AIXSocketOptions_setQuickAck0
+(JNIEnv *env, jobject unused, jint fd, jboolean on) {
+    int optval;
+    int rv;
+    optval = (on ? 1 : 0);
+    rv = setsockopt(fd, SOL_SOCKET, TCP_NODELAYACK, &optval, sizeof (optval));
+    handleError(env, rv, "set option TCP_NODELAYACK failed");
+}
+
+/*
+ * Class:     jdk_net_AIXSocketOptions
+ * Method:    getQuickAck
+ * Signature: (I)Z;
+ */
+JNIEXPORT jboolean JNICALL Java_jdk_net_AIXSocketOptions_getQuickAck0
+(JNIEnv *env, jobject unused, jint fd) {
+    int on;
+    socklen_t sz = sizeof (on);
+    int rv = getsockopt(fd, SOL_SOCKET, TCP_NODELAYACK, &on, &sz);
+    handleError(env, rv, "get option TCP_NODELAYACK failed");
+    return on != 0;
+}
+
+/*
+ * Class:     jdk_net_AIXSocketOptions
+ * Method:    quickAckSupported
+ * Signature: ()Z
+ */
+JNIEXPORT jboolean JNICALL Java_jdk_net_AIXSocketOptions_quickAckSupported0
+(JNIEnv *env, jobject unused) {
+    return socketOptionSupported(SOL_SOCKET, TCP_NODELAYACK);
+}
+
+/*
+ * Class:     jdk_net_AIXSocketOptions
+ * Method:    getSoPeerCred0
+ * Signature: (I)L
+ */
+JNIEXPORT jlong JNICALL Java_jdk_net_AIXSocketOptions_getSoPeerCred0
+  (JNIEnv *env, jclass clazz, jint fd) {
+
+    int rv;
+    struct peercred_struct cred_info;
+    socklen_t len = sizeof(cred_info);
+
+    if ((rv=getsockopt(fd, SOL_SOCKET, SO_PEERID, &cred_info, &len)) < 0) {
+        handleError(env, rv, "get  failed");
```

Comment:
Is the double space after "get  " intended here, or is something missing from the error message?

Reply:
It is intended

Reply:
I'm curious, what is the rationale?
By comparison the Linux version has: 
```
        handleError(env, rv, "get SO_PEERCRED failed");
```

Reply:
On linux it is implemented from [attachListener_linux.cpp](https://github.com/openjdk/jdk/blob/master/src/hotspot/os/linux/attachListener_linux.cpp#L361) and on aix it is from [attachListener_aix.cpp](https://github.com/openjdk/jdk/blob/master/src/hotspot/os/aix/attachListener_aix.cpp#L392)
In both the cases when rv == -1 , it fails to get this socket option.

Reply:
Hello Varada, so the Linux version uses `SO_PEERCRED` socket option and when it fails, the error message says "get SO_PEERCRED failed". The AIX version uses `SO_PEERID` and if it fails, the current proposed form in this PR will report the error message as "get  failed" which wouldn't be too informative.

I'm guessing that the reason you didn't include the socket option in the error message is perhaps because the native socket option (on AIX) is `SO_PEERID` but this method is named `getSoPeerCred`. So it might be confusing to report `SO_PEERID` in the error message. Maybe you could change the message to just "getSoPeerCred failed"  instead?

Reply:
The method name `getSoPeerCred` is the one used in ExtendedSocketOptions.java, which we are overriding here.
The supported socket option on AIX for is  `SO_PEERID` . I will add `"get SO_PEERID failed"`

Reply:
For consistency reasons I think I would prefer the message to be "get SO_PEERID failed". All the other methods in this file seem to use the native option name, so let's keep things consistent.